cmd/zpool/zpool_main.c

```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
 }
 
 /*
- * zpool iostat [-gLPv] [-T d|u] [pool] ... [interval [count]]
+ * Return stat flags that are supported by all pools.  "data" should be
+ * initialized to all 0xFFs before running.  It will get ANDed down until
+ * only the flags that are supported on all pools remain.
+ */
+static int
+__get_stat_flags(zpool_handle_t *zhp, void *data)
+{
+	uint64_t *mask = data;
+	nvlist_t *config, *nvroot;
+	uint64_t flags;
+
+	config = zpool_get_config(zhp, NULL);
+	verify(nvlist_lookup_nvlist(config, ZPOOL_CONFIG_VDEV_TREE,
+	    &nvroot) == 0);
+
+	/* What stats does our zfs module support? */
+	if (nvlist_lookup_uint64(nvroot, ZPOOL_CONFIG_VDEV_STATS_FLAGS,
+		&flags) != 0) {
+		/* Probably an old version of the zfs module.  Use defaults. */
+		flags = BIT(VS_DEFAULT);
+	}
+
+	*mask = *mask & flags;
+	return (0);
+}
+
+/*
+ * Newer versions of the zfs module report back a bitmask of all the stats
+ * it supports.  Return the bitmask of stats that are supported on all pools.
+ */
+static uint64_t
+get_stat_flags(zpool_list_t *list)
+{
+	uint64_t mask = -1; /* Start off with all 0xFFs */
+
+	/*
+	 * __get_stat_flags() will lop off bits from "mask" until only the
+	 * flags that are supported on all pools remain.
+	 */
+	pool_list_iter(list, B_FALSE, __get_stat_flags, &mask);
+	return (mask);
+}
+
+/* for_each_vdev() function prototype */
+typedef int (*pool_vdev_iter_f)(zpool_handle_t *, nvlist_t *, void *);
+
+static int
+__for_each_vdev(zpool_handle_t *zhp, nvlist_t *nv, pool_vdev_iter_f func,
+		void *data)
+{
+	nvlist_t **child;
+	uint_t c, children;
+	int ret = 0;
+	int i;
+	char *type;
+
+	const char *list[] = {
+		ZPOOL_CONFIG_SPARES,
+		ZPOOL_CONFIG_L2CACHE,
+		ZPOOL_CONFIG_CHILDREN
+	};
+
+	for (i = 0; i < ARRAY_SIZE(list); i++) {
+		if (nvlist_lookup_nvlist_array(nv, list[i],
+					&child, &children) == 0) {
+			for (c = 0; c < children; c++) {
+				uint64_t ishole = B_FALSE;
+
+				(void) nvlist_lookup_uint64(child[c],
+					ZPOOL_CONFIG_IS_HOLE,
+					&ishole);
+
+				if (ishole)
+					continue;
+
+				ret |= __for_each_vdev(zhp, child[c], func,
+							data);
+			}
+		}
+	}
+
+	if (nvlist_lookup_string(nv, ZPOOL_CONFIG_TYPE, &type) != 0)
+		goto end;
+
+	/* Don't run our function on root vdevs */
+	if (strcmp(type, VDEV_TYPE_ROOT) != 0) {
+		ret += func(zhp, nv, data);
+	}
+
+end:
+	return (ret);
+}
+
+/*
+ * This is the equivalent of for_all_pools() for vdevs.  It iterates thorough
```

Comment:
You mean `for_each_pool()`, and this iterate logic should be part of `cmd/zpool/zpool_iter.c`